File: app/src/main/java/dev/ragnarok/fenrir/api/services/IFaveService.java
package dev.ragnarok.fenrir.api.services;

import dev.ragnarok.fenrir.api.model.FaveLinkDto;
import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiArticle;
import dev.ragnarok.fenrir.api.model.VKApiPhoto;
import dev.ragnarok.fenrir.api.model.VkApiAttachments;
import dev.ragnarok.fenrir.api.model.response.BaseResponse;
import dev.ragnarok.fenrir.api.model.response.FavePageResponse;
import dev.ragnarok.fenrir.api.model.response.FavePostsResponse;
import io.reactivex.rxjava3.core.Single;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.POST;

public interface IFaveService {

    @FormUrlEncoded
    @POST("fave.getPages")
    Single<BaseResponse<Items<FavePageResponse>>> getPages(@Field("offset") Integer offset,
                                                           @Field("count") Integer count,
                                                           @Field("type") String type,
                                                           @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.get")
    Single<BaseResponse<Items<VkApiAttachments.Entry>>> getVideos(@Field("offset") Integer offset,
                                                                  @Field("count") Integer count,
                                                                  @Field("item_type") String item_type,
                                                                  @Field("extended") Integer extended,
                                                                  @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.get")
    Single<BaseResponse<Items<VkApiAttachments.Entry>>> getArticles(@Field("offset") Integer offset,
                                                                    @Field("count") Integer count,
                                                                    @Field("item_type") String item_type,
                                                                    @Field("extended") Integer extended,
                                                                    @Field("fields") String fields);

    @FormUrlEncoded
    @POST("articles.getOwnerPublished")
    Single<BaseResponse<Items<VKApiArticle>>> getOwnerPublishedArticles(@Field("owner_id") Integer owner_id,
                                                                        @Field("offset") Integer offset,
                                                                        @Field("count") Integer count,
                                                                        @Field("sort_by") String sort_by,
                                                                        @Field("extended") Integer extended,
                                                                        @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.get")
    Single<BaseResponse<FavePostsResponse>> getPosts(@Field("offset") Integer offset,
                                                     @Field("count") Integer count,
                                                     @Field("item_type") String item_type,
                                                     @Field("extended") Integer extended,
                                                     @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.get")
    Single<BaseResponse<Items<FaveLinkDto>>> getLinks(@Field("offset") Integer offset,
                                                      @Field("count") Integer count,
                                                      @Field("item_type") String item_type,
                                                      @Field("extended") Integer extended,
                                                      @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.get")
    Single<BaseResponse<Items<VkApiAttachments.Entry>>> getProducts(@Field("offset") Integer offset,
                                                                    @Field("count") Integer count,
                                                                    @Field("item_type") String item_type,
                                                                    @Field("extended") Integer extended,
                                                                    @Field("fields") String fields);

    @FormUrlEncoded
    @POST("fave.getPhotos")
    Single<BaseResponse<Items<VKApiPhoto>>> getPhotos(@Field("offset") Integer offset,
                                                      @Field("count") Integer count);

    @FormUrlEncoded
    @POST("fave.addLink")
    Single<BaseResponse<Integer>> addLink(@Field("link") String link);

    @FormUrlEncoded
    @POST("fave.addPage")
    Single<BaseResponse<Integer>> addPage(@Field("user_id") Integer userId,
                                          @Field("group_id") Integer groupId);

    @FormUrlEncoded
    @POST("fave.addVideo")
    Single<BaseResponse<Integer>> addVideo(@Field("owner_id") Integer owner_id,
                                           @Field("id") Integer id,
                                           @Field("access_key") String access_key);

    @FormUrlEncoded
    @POST("fave.addArticle")
    Single<BaseResponse<Integer>> addArticle(@Field("url") String url);

    @FormUrlEncoded
    @POST("fave.addProduct")
    Single<BaseResponse<Integer>> addProduct(@Field("id") int id,
                                             @Field("owner_id") int owner_id,
                                             @Field("access_key") String access_key);

    @FormUrlEncoded
    @POST("fave.addPost")
    Single<BaseResponse<Integer>> addPost(@Field("owner_id") Integer owner_id,
                                          @Field("id") Integer id,
                                          @Field("access_key") String access_key);

    //https://vk.com/dev/fave.removePage
    @FormUrlEncoded
    @POST("fave.removePage")
    Single<BaseResponse<Integer>> removePage(@Field("user_id") Integer userId,
                                             @Field("group_id") Integer groupId);

    @FormUrlEncoded
    @POST("fave.removeLink")
    Single<BaseResponse<Integer>> removeLink(@Field("link_id") String linkId);

    @FormUrlEncoded
    @POST("fave.removeArticle")
    Single<BaseResponse<Integer>> removeArticle(@Field("owner_id") Integer owner_id,
                                                @Field("article_id") Integer article_id);

    @FormUrlEncoded
    @POST("fave.removeProduct")
    Single<BaseResponse<Integer>> removeProduct(@Field("id") Integer id,
                                                @Field("owner_id") Integer owner_id);

    @FormUrlEncoded
    @POST("fave.removePost")
    Single<BaseResponse<Integer>> removePost(@Field("owner_id") Integer owner_id,
                                             @Field("id") Integer id);

    @FormUrlEncoded
    @POST("fave.removeVideo")
    Single<BaseResponse<Integer>> removeVideo(@Field("owner_id") Integer owner_id,
                                              @Field("id") Integer id);

    @FormUrlEncoded
    @POST("execute")
    Single<BaseResponse<Integer>> pushFirst(@Field("code") String code,
                                            @Field("owner_id") int ownerId);

}

File: app/src/main/kotlin/dev/ragnarok/fenrir/api/services/IFaveService.kt
package dev.ragnarok.fenrir.api.services

import dev.ragnarok.fenrir.api.model.*
import dev.ragnarok.fenrir.api.model.response.BaseResponse
import dev.ragnarok.fenrir.api.model.response.FavePageResponse
import dev.ragnarok.fenrir.api.model.response.FavePostsResponse
import io.reactivex.rxjava3.core.Single
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.POST

interface IFaveService {
    @FormUrlEncoded
    @POST("fave.getPages")
    fun getPages(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("type") type: String?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<FavePageResponse>>>

    @FormUrlEncoded
    @POST("fave.get")
    fun getVideos(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("item_type") item_type: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<VKApiAttachments.Entry>>>

    @FormUrlEncoded
    @POST("fave.get")
    fun getArticles(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("item_type") item_type: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<VKApiAttachments.Entry>>>

    @FormUrlEncoded
    @POST("articles.getOwnerPublished")
    fun getOwnerPublishedArticles(
        @Field("owner_id") owner_id: Int?,
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("sort_by") sort_by: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<VKApiArticle>>>

    @FormUrlEncoded
    @POST("fave.get")
    fun getPosts(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("item_type") item_type: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<FavePostsResponse>>

    @FormUrlEncoded
    @POST("fave.get")
    fun getLinks(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("item_type") item_type: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<FaveLinkDto>>>

    @FormUrlEncoded
    @POST("fave.get")
    fun getProducts(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?,
        @Field("item_type") item_type: String?,
        @Field("extended") extended: Int?,
        @Field("fields") fields: String?
    ): Single<BaseResponse<Items<VKApiAttachments.Entry>>>

    @FormUrlEncoded
    @POST("fave.getPhotos")
    fun getPhotos(
        @Field("offset") offset: Int?,
        @Field("count") count: Int?
    ): Single<BaseResponse<Items<VKApiPhoto>>>

    @FormUrlEncoded
    @POST("fave.addLink")
    fun addLink(@Field("link") link: String?): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.addPage")
    fun addPage(
        @Field("user_id") userId: Int?,
        @Field("group_id") groupId: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.addVideo")
    fun addVideo(
        @Field("owner_id") owner_id: Int?,
        @Field("id") id: Int?,
        @Field("access_key") access_key: String?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.addArticle")
    fun addArticle(@Field("url") url: String?): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.addProduct")
    fun addProduct(
        @Field("id") id: Int,
        @Field("owner_id") owner_id: Int,
        @Field("access_key") access_key: String?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.addPost")
    fun addPost(
        @Field("owner_id") owner_id: Int?,
        @Field("id") id: Int?,
        @Field("access_key") access_key: String?
    ): Single<BaseResponse<Int>>

    //https://vk.com/dev/fave.removePage
    @FormUrlEncoded
    @POST("fave.removePage")
    fun removePage(
        @Field("user_id") userId: Int?,
        @Field("group_id") groupId: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.removeLink")
    fun removeLink(@Field("link_id") linkId: String?): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.removeArticle")
    fun removeArticle(
        @Field("owner_id") owner_id: Int?,
        @Field("article_id") article_id: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.removeProduct")
    fun removeProduct(
        @Field("id") id: Int?,
        @Field("owner_id") owner_id: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.removePost")
    fun removePost(
        @Field("owner_id") owner_id: Int?,
        @Field("id") id: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("fave.removeVideo")
    fun removeVideo(
        @Field("owner_id") owner_id: Int?,
        @Field("id") id: Int?
    ): Single<BaseResponse<Int>>

    @FormUrlEncoded
    @POST("execute")
    fun pushFirst(
        @Field("code") code: String?,
        @Field("owner_id") ownerId: Int
    ): Single<BaseResponse<Int>>
}